go/tasks/pluginmachinery/google/gke_token_source.go

```diff
@@ -0,0 +1,212 @@
+	createTokenResponse, err := serviceAccounts.CreateToken(ctx, identity.K8sServiceAccount, &v1.TokenRequest{
+		Spec: v1.TokenRequestSpec{
+			Audiences: []string{
+				m.identityNamespace,
+			},
+		},
+	}, metav1.CreateOptions{})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create k8s service account token")
+	}
+
+	return createTokenResponse.Status.Token, nil
+}
+
+func (m *gkeTokenSource) getGcpServiceAccount(ctx context.Context, identity Identity) (string, error) {
+	serviceAccounts := m.kubeClient.CoreV1().ServiceAccounts(identity.K8sNamespace)
+
+	serviceAccountResponse, err := serviceAccounts.Get(ctx, identity.K8sServiceAccount, metav1.GetOptions{})
+
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create k8s service account token")
+	}
+
+	for key, value := range serviceAccountResponse.Annotations {
+		if key == gcpServiceAccountAnnotationKey {
+			return value, nil
+		}
+	}
+
+	return "", errors.Errorf(
+		"[%v] annotation doesn't exist on k8s service account [%v/%v], read more at %v",
+		gcpServiceAccountAnnotationKey,
+		identity.K8sNamespace,
+		identity.K8sServiceAccount,
+		workflowIdentityDocURL)
+}
+
+func (m gkeTokenSource) GetTokenSource(ctx context.Context, identity Identity) (oauth2.TokenSource, error) {
+	if identity.K8sServiceAccount == "" {
+		identity.K8sServiceAccount = "default"
+	}
+
+	if cachedToken, ok := m.getCachedToken(ctx, identity); ok {
+		return oauth2.StaticTokenSource(cachedToken), nil
+	}
+
+	// when tokens expire, of we hit a miss, singleflight will do a most one request per SA
+	value, err := m.singleflight.Do(identity.K8sNamespace+"/"+identity.K8sServiceAccount, func() (interface{}, error) {
+		k8sServiceAccountToken, err := m.getK8sServiceAccountToken(ctx, identity)
+
+		if err != nil {
+			return nil, err
+		}
+
+		gcpServiceAccount, err := m.getGcpServiceAccount(ctx, identity)
+
+		if err != nil {
+			return nil, err
+		}
+
+		token, err := ExchangeToken(ctx, StsRequest{
+			SubjectToken:           k8sServiceAccountToken,
+			Scope:                  m.scope,
+			ServiceAccount:         gcpServiceAccount,
+			IdentityNamespace:      m.identityNamespace,
+			iamCredentialsEndpoint: m.iamCredentialsEndpoint,
+			federatedTokenEndpoint: m.federatedTokenEndpoint,
+		})
+
+		if err != nil {
+			return nil, err
+		}
+
+		m.tokens.Store(identity, token)
+
+		return token, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return oauth2.StaticTokenSource(value.(*oauth2.Token)), nil
+}
+
+func getKubeClient(kubeConfigPath string, kubeConfig KubeClientConfig) (*kubernetes.Clientset, error) {
+	var kubecfg *rest.Config
+	var err error
+	if kubeConfigPath != "" {
+		kubeConfigPath := os.ExpandEnv(kubeConfigPath)
+		kubecfg, err = clientcmd.BuildConfigFromFlags("", kubeConfigPath)
+		if err != nil {
+			return nil, errors.Wrapf(err, "Error building kubeconfig")
+		}
+	} else {
+		kubecfg, err = rest.InClusterConfig()
```

Comment:
if you want to do incluster config, better to rely on the kubeclient, that is part of the TaskExecutionContext?